src/schema.ts

```diff
@@ -173,7 +181,281 @@ class SchemaDefinition implements SchemaDefinitionType {
             updateAdjList(adjacencyList, level, removed);
         }
 
-        return levels.map((level) => level.map((modelName) => this.schema[modelName]));
+        return levels;
+    }
+
+    /**
+     * a list of the properties associate with this class or parents of this class
+     * @type {Array.<PropertyDefinition>}
+     */
+    getProperties(modelName: string): Record<string, PropertyDefinition> {
+        const model = this.get(modelName);
+        let properties: Record<string, PropertyDefinition> = { ...model.properties };
+
+        for (const parent of this.ancestors(modelName)) {
+            properties = { ...this.getProperties(parent), ...properties }; // properties of the same name are taken from the lowest model
+        }
+        return properties;
+    }
+
+    requiredProperties(modelName: string): string[] {
+        return Object.values(this.getProperties(modelName))
+            .filter((propDefn) => propDefn.mandatory)
+            .map((propDefn) => propDefn.name);
+    }
+
+    optionalProperties(modelName: string): string[] {
+        return Object.values(this.getProperties(modelName))
+            .filter((propDefn) => !propDefn.mandatory)
+            .map((propDefn) => propDefn.name);
+    }
+
+    getProperty(modelName: string, property: string): PropertyDefinition {
+        return this.getProperties(modelName)[property];
+    }
+
+    hasProperty(modelName: string, property: string): boolean {
+        return Boolean(this.getProperty(modelName, property));
+    }
+
+    inheritsFrom(modelName: string, parent: string): boolean {
+        return this.ancestors(modelName).includes(parent);
+    }
+
+    /**
+     * Returns the list of properties for the Class.active index. This is
+     * expected to represent a unique constraint on a combination of these properties
+     *
+     * @returns list of properties which belong to the index
+     */
+    activeProperties(modelName: string): string[] | null {
+        const model = this.get(modelName);
+
+        for (const index of model.indices) {
+            if (index.name === `${model.name}.active`) {
+                return index.properties;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * the list of parent class names which this class inherits from
+     */
+    ancestors(modelName: string): string[] {
+        const model = this.get(modelName);
+        const parents: string[] = [];
+
+        for (const parent of model.inherits) {
+            parents.push(parent);
+            parents.push(...this.ancestors(parent));
+        }
+        return parents;
+    }
+
+    children(modelName: string): string[] {
```

Comment:
I would consider renaming this it's this method is going to be used with the gui at all. maybe could use the same name (successors) that networkx does?

Reply:
what kind of issues would it cause with the client? 

technically this only returns the immediate successors of a given model, whereas successors itself is really more equivalent to the descendants function since it is just any node along a path from x to y (in traditional graph theory, not how networkx uses the term). It might be confusing to have both successors and descendants and have them mean separate things

Reply:
I guess I'm used to networkx which has both [successors](https://networkx.org/documentation/stable/reference/classes/generated/networkx.DiGraph.successors.html) and [descendants](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.dag.descendants.html) and that's how they use them. 
`children` is a reserved prop name in react so if I saw this function called within a component I might be expecting it to return JSX

Reply:
I don't like either option. I am going to leave this as is for now and I will think more on it. We have a few PRs before release still